client/gutenberg/editor/calypsoify-iframe.tsx

```diff
@@ -38,28 +37,57 @@ import { getEditorPostId } from 'state/ui/editor/selectors';
  */
 import './style.scss';
 
-class CalypsoifyIframe extends Component {
-	static propTypes = {
-		postId: PropTypes.number,
-		postType: PropTypes.string,
-		duplicatePostId: PropTypes.number,
-		pressThis: PropTypes.object,
-	};
+interface Props {
+	duplicatePostId: number;
+	postId: number;
+	postType: string;
+	pressThis: any;
+	siteAdminUrl: string;
+}
+
+type ConnectedProps = ReturnType< typeof mapStateToProps > & typeof mapDispatchToProps;
+
+interface State {
+	allowedTypes?: any;
+	classicBlockEditorId?: any;
+	editedPost?: any;
+	gallery?: any;
+	isIframeLoaded: boolean;
+	isMediaModalVisible: boolean;
+	isPreviewVisible: boolean;
+	multiple?: any;
+	postUrl?: string;
+	previewUrl: string;
+}
+
+enum WindowActions {
+	Loaded = 'loaded',
+	ClassicBlockOpenMediaModel = 'classicBlockOpenMediaModal',
+}
+
+enum EditorActions {
+	GoToAllPosts = 'goToAllPosts',
+	OpenMediaModal = 'openMediaModal',
+	OpenRevisions = 'openRevisions',
+	PreviewPost = 'previewPost',
+	SetDraftId = 'draftIdSet',
+	TrashPost = 'trashPost',
+}
 
-	state = {
-		classicBlockEditorId: null,
+class CalypsoifyIframe extends Component< Props & ConnectedProps, State > {
+	state: State = {
 		isMediaModalVisible: false,
 		isIframeLoaded: false,
 		isPreviewVisible: false,
 		previewUrl: 'about:blank',
-		postUrl: null,
-		editedPost: null,
 	};
 
-	constructor( props ) {
-		super( props );
-		this.iframeRef = React.createRef();
-		this.mediaSelectPort = null;
+	iframeRef: React.RefObject< HTMLIFrameElement > = React.createRef();
+	iframePort: MessagePort | null = null;
+	mediaSelectPort: MessagePort | null = null;
+
+	constructor( ...args ) {
+		super( ...args );
 		MediaStore.on( 'change', this.updateImageBlocks );
```

Comment:
This looks like a pre-existing bug. The `MediaStore.on( 'change' )` subscription should be created in `componentDidMount` (not in constructor) and removed in `componentWillUnmount` (not unsubscribed at all currently). Then the constructor can disappear.

Reply:
Great catch. For the missing warrant, I believe that we can end up calling the `constructor()` even in cases where the component never mounts, likewise we might run into situations where `constructor()` is called multiple times before mounting. Sound right?

I'll try and fix this bug outside of this PR but fold in the changes here. That's a behavioral change I'd like to isolate.

Reply:
> For the missing warrant, I believe that we can end up calling the constructor() even in cases where the component never mounts, likewise we might run into situations where constructor() is called multiple times before mounting.

Yes, with React Fiber, error boundaries and concurrent mode, it's increasingly common to perform some work (constructing a component), then throw it away for some reason (error thrown from render and caught by error boundary, high-priority concurrent update preempting an already running one, suspending a render) and then restart it again.

`constructor` doesn't have an associated destructor, so acquiring resources in it can lead to leaks.